MySensors binding added

tobof opened the pull request:
New binding that adds support for the MySensors protocol / network.

https://www.mysensors.org

Discussion about the binding development can be found here: https://forum.mysensors.org/topic/1598/openhab-2-0-binding

kubawolanin commented:
Quick question - would this binding work with [ESP8266 Gateway](https://www.mysensors.org/build/esp8266_gateway)? I'm pretty newbie in MySensors and would like to play around with your binding :slightly_smiling_face: 
Thanks!

tobof commented:
@kubawolanin Yes it works with the Gateway you mentioned. Just use the "ethernet gateway" configuration described in the wiki.

fab33 commented:
@kaikreuzer can you merge this branch with openhab2-addon ?

kaikreuzer commented:
@tobof There seems to be still development going on on that branch. I'll definitely wait until it is squashed to a single commit and the code respects all the [guidelines](http://docs.openhab.org/developers/development/guidelines.html). Meanwhile, I have plenty of other bindings in the queue to review...

kaikreuzer commented:
> Meanwhile, I have plenty of other bindings in the queue to review...

Actually, I have just catched up and this one would be the next to be reviewed :-)

So @tobof, please be so kind to check your code against the guidelines and ping me when you think it is good to go - thanks!

hakan42 commented on addons/binding/org.openhab.binding.mysensors/pom.xml:
```diff
@@ -0,0 +1,19 @@
+
+  <groupId>org.openhab.binding</groupId>
+  <artifactId>org.openhab.binding.mysensors</artifactId>
+  <version>2.0.0-SNAPSHOT</version>
```
Please remove groupId and version from the pom, that way it will be inherited from the parent.

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/exception/NoMoreIdsException.java:
```diff
@@ -0,0 +1,23 @@
+package org.openhab.binding.mysensors.internal.exception;
+
+/**
+ * Exception occures if no more ids are available.
```
Typo - "occurs" :)

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/handler/MySensorsBridgeHandler.java:
```diff
@@ -0,0 +1,282 @@
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
```
please remove empty handleCommand() if not necessary.

tobof replied:
The abstract method handleCommand() needs to be implemented. That's what eclipse tells me. :)

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/handler/MySensorsThingHandler.java:
```diff
@@ -0,0 +1,413 @@
+     */
+    private void handleIncomingMessageEvent(MySensorsMessage msg) {
+        // Am I the all knowing node that receives all messages?
+        if (nodeId == 999 && childId == 999) {
```
Could you please create constants for those "all knowing node and child ids" ?

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/protocol/message/MySensorsMessage.java:
```diff
@@ -0,0 +1,246 @@
+     * Write message to DEBUG.
+     */
+    public void printDebug() {
+        logger.debug(String.format("nodeId: %d, childId: %d, msgType: %d, ack: %d, revert: %b, subType: %d ,msg: %s",
```
could this be replaced by logger.debug(getDebugInfo())? the two strings look very similiar.

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/protocol/message/MySensorsMessage.java:
```diff
@@ -0,0 +1,246 @@
+    public boolean isIConfigMessage() {
+        boolean ret = false;
+
+        if (childId == 0 || childId == 255) {
```
Please create constants for magic childIds (in this case 255?)

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/protocol/serial/MySensorsSerialConnection.java:
```diff
@@ -0,0 +1,199 @@
+        try {
+            String[] namePart = devName.split("/");
+            File lockFile = new File("/var/lock/LCK.." + namePart[namePart.length - 1]);
+            if (lockFile.exists()) {
```
The path /var/lock/... only exists on Unix, right? Does this serial interface library work on Windows or MacOS?

hakan42 commented on addons/binding/pom.xml:
```diff
@@ -65,6 +65,7 @@
     <module>org.openhab.binding.toon</module>
     <module>org.openhab.binding.urtsi</module>
     <module>org.openhab.binding.vitotronic</module>
+    <module>org.openhab.binding.mysensors</module>
```
Please keep list of bindings in pom alphabetically sorted.

hakan42 commented on addons/binding/pom.xml~:
```diff
@@ -0,0 +1,49 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Please remove accidentally committed backup file

tobof commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/handler/MySensorsThingHandler.java:
```diff
@@ -0,0 +1,413 @@
+    }
+
+    @Override
+    public void handleUpdate(ChannelUID channelUID, org.eclipse.smarthome.core.types.State newState) {
```
This method is inherited from BaseThingHandler and must be implemented.

tobof commented:
Thank you for your suggestions @hakan42 !
I've made some changes and I'm open to new suggestions! :)

fab33 commented:
Thanks a lot !

kaikreuzer commented:
@tobof Once you are done and feel like it is ready to be reviewed/merged, please make sure to squash your commits first - 184 is an awful number of commits 😲

ThomDietrich commented:
Out of interest: Why are many commits not better for the review? I feel like they would help understand the result and a squash should be done only right before the merge. But maybe that's just my thinking, not being involved in the review process and all... :smiley:

hakan42 commented on addons/binding/org.openhab.binding.mysensors/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,33 @@
+Import-Package: 
+ com.google.common.base,
+ com.google.common.collect,
+ com.google.gson;version="2.2.4",
```
Do you need that specific version of gson? If not, please remove version constraint.

hakan42 commented on addons/binding/org.openhab.binding.mysensors/.gitignore:
```diff
@@ -0,0 +1 @@
+/*.cached
```
Does something in your project create the *.cached files / directories? If not, please remove project specific gitignore file

hakan42 commented on addons/binding/org.openhab.binding.mysensors/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,1476 @@
+			<parameter name="baudRate" type="integer" required="false" min="300" max="250000">
+				<label>Baudrate</label>
+				<description>Baudrate for serial port
+                </description>
```
Once remarked for this, file, happens many times: Please use tabs instead of spaces for indenting.

kaikreuzer commented:
If "many" would mean 10, this could be better, yes.
But having 184, many of which are not signed-off, others being merge commits, is simply not helpful.

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/MySensorsBindingConstants.java:
```diff
@@ -0,0 +1,400 @@
+     * Mapping MySensors subtypes to channels.
+     */
+    public final static Map<Number, String> CHANNEL_MAP = new HashMap<Number, String>() {
+        /**
```
Remove empty Javadoc please

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/factory/MySensorsCacheFactory.java:
```diff
@@ -0,0 +1,138 @@
+
+    private static MySensorsCacheFactory singleton = null;
+
+    private static final String CACHE_BASE_PATH = "./mysensors/cache";
```
Where would this cache be written in a production environment? I believe it should belong into userdata, but not into the current working directory of the process. To which you might even not have write access....

tobof replied:
It is written to userdata. In my production environment it is written to: _/var/lib/openhab2/mysensors/cache/given_ids.cached_

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/protocol/MySensorsBridgeConnection.java:
```diff
@@ -0,0 +1,595 @@
+    public void broadCastEvent(MySensorsStatusUpdateEvent event) {
+        synchronized (updateListeners) {
+            for (MySensorsUpdateListener mySensorsEventListener : updateListeners) {
+                logger.trace("Broadcasting event to: " + mySensorsEventListener);
```
Please use parametrized logging instead of string concatenation.

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/protocol/MySensorsBridgeConnection.java:
```diff
@@ -0,0 +1,595 @@
+                    iterator.remove();
+                    // logger.debug("Message removed: " + msg.getDebugInfo());
+                } else {
+                    logger.debug("Message NOT removed: " + msg.getDebugInfo());
```
Please use parametrized logging instead of string concatenation.

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/protocol/MySensorsBridgeConnection.java:
```diff
@@ -0,0 +1,595 @@
+     * @param flag true if the connection should be stopped.
+     */
+    public void requestDisconnection(boolean flag) {
+        logger.debug("Request disconnection flag setted to: " + flag);
```
Please use parametrized logging instead of string concatenation.

tobof commented:
@kaikreuzer I'm done preparing, but I need some help to squash / rebase. I've read a few tutorials and tried it with the local repository. My main problem: If I start a rebase and start here: f69bf36153cc7e0db67195dc68ae602e981d70c1 (first commit). The list ist about 700 commits long (only a few from me). Most tutorials state: "Don't do a rebase on a remote repository!" Could you please give me a hint?

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/protocol/MySensorsWriter.java:
```diff
@@ -0,0 +1,163 @@
+                                            + MySensorsBindingConstants.MYSENSORS_RETRY_TIMES[msg.getRetries() - 1]);
+                                    mysCon.addMySensorsOutboundMessage(msg);
+                                } else {
+                                    logger.warn("NO ACK from nodeId: " + msg.getNodeId());
```
Please use parametrized logging instead of string concatenation.

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/protocol/ip/MySensorsIpConnection.java:
```diff
@@ -0,0 +1,101 @@
+
+                ret = startReaderWriterThread(mysConReader, mysConWriter);
+            } catch (UnknownHostException e) {
+                logger.error("Error while trying to connect to: " + ipAddress + ":" + tcpPort);
```
Please use parametrized logging instead of string concatenation.

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/protocol/ip/MySensorsIpWriter.java:
```diff
@@ -0,0 +1,37 @@
+            outs = new PrintWriter(outStream);
+        } catch (IOException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
```
Please use slf4j instead of e.printStackTrace() 😈

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/sensors/MySensorsDeviceManager.java:
```diff
@@ -0,0 +1,217 @@
+    @SuppressWarnings("unused")
+    private void checkChildFound(MySensorsMessage msg) {
+        synchronized (nodeMap) {
+            if (msg.childId != 255 && !nodeMap.containsKey(msg.childId)) {
```
Magic Child ID 255. I believe we have constants for that?

hakan42 commented on addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/sensors/MySensorsDeviceManager.java:
```diff
@@ -0,0 +1,217 @@
+        int newId = 0;
+        try {
+            newId = reserveId();
+            MySensorsMessage newMsg = new MySensorsMessage(255, 255, 3, 0, false, 4, newId + "");
```
More magic numbers, please replace with constants where appropiate.

hakan42 commented:
@tobof , actually, you could create a new clone on your computer, then locally delete your branch ("git branch -D your-name), check out master again, create a new branch with the same name ("git checkout -b your-name"), add all your files from your backup directory, then do one big commit and the force-push that branch.

That way, this PR will automatically have only the one commit. I prefer this way when master and my base have diverged too much.

tobof commented:
@hakan42 Thank you for the suggestion! Currently the PR is from my master branch. If I delete my master branch (is that even possible?) are you sure this will preserve the PR. 

The ideal solution in my opinion is:

MySensors branch with all commits
--> squashed in my master branch to one commit
--> PR to the official openhab2-addons master branch.

This way I would preserve the original history and all contributors in the MySensors branch and there's only one commit in the PR to openhab2-addons. But how to achieve that AND preserve this PR? :)

hakan42 commented:
As far as I understand, you would develop in your own master branch of your own repo as you want (perfectly o.k.), and then create a seperate branch only for the PR. Effectively, you are "exporting" your complete changeset as one huge commit.

I believe preserving this PR will be a little bit hard. you could rename your master to "real-master" (see for example https://gist.github.com/lttlrck/9628955 ), create a new "master", add everything from "real-master" in one commit, and force-push that.

As soon as this PR is merged, you could rename your "real-master" back to master and return to normal development.

A bit dangerous, all that branch hacking. Better keep good backups of your repositories before trying any of that stupid stuff I mentioned 😄

ThomDietrich commented:
> As far as I understand, you would develop in your own master branch

That's actually something you should never do. Always develop in a feature branch.

Anyways. @tobof you have two options: You can [squash all your commits](https://stackoverflow.com/questions/6934752/combining-multiple-commits-before-pushing-in-git) or simply [re-commit everything](http://stackoverflow.com/a/8214049/6654920). Whatever you do, you can check the results before pushing, so nothing should go wrong. To be safe create a copy of your local project. If anything happens you can restore from there.

hakan42 commented:
@ThomDietrich you are right with "never develop in master", I was just trying to come up with an idea to rescue this PR 😄 

That said, if one of my team members at $WORPLACE would have dared to develop in master.... I am the bastard operator who takes away push privileges 😈

tobof commented:
Ok, so I've reduced the number of commits to one, thank you for all the suggestions. I've tried @ThomDietrich suggestions, but that didn't work out, because I squashed all upstream commits into my commit. I suppose I've done something wrong. After that I got along with @hakan42 suggestion.

@hakan42 : I've implemented all your request for changes!

tobof commented:
@kaikreuzer Pls remove awaiting feedback flag, I'm done. Thank you! :-)

kaikreuzer commented:
Thanks for the ping.
The code will need some further review, though.
While waiting, you might want to check it yourself against the [coding guidelines](http://docs.openhab.org/developers/development/guidelines.html), at a first glance I saw problems with A.2 (old headers), B.4 and B.5. Please also be so kind to update the version to 2.1.0 already - thanks!

tobof commented:
@kaikreuzer Thank you for the hint! I've corrected A.2, B.4 and B.5.

According other guidelines: D.2 may be a problem, because threads are used within the binding.

mcinnes01 commented:
Hi just wondered if there is any update on this pull request and idea when it will be released?

hakan42 commented:
@tobof , do you want to abandon this binding or continue development in a new PR? Would be a shame if you abandoned it, you seem to have lots of users on the mysensors.org forums 😄

tobof commented:
@hakan42 We've done a major refactoring of the binding (credits go to @andreacioni) and I'm cleaning up my fork atm. 
I'll open a new PR after checking the code against the coding guidelines again and after I'm finished struggling with the static code analysis tool. :)
